### PR TITLE
Mcm Allocation and Controllers

### PR DESCRIPTION
**Pull request by Herpderk**
Changelogs 

ADCS
- Magnetic torque coil voltage allocator 
- Controllers implementation
- Mode management implementation 

CD&H
- Availability and clearing capabilities to DH
- Scheduler trim and documentation

TODO:
1. hardware sanity check

### CODE REVIEW

File: flight/apps/adcs/mcm.py [hunk 1]
```diff
@@ -7,4 +7,236 @@
...
+        0.0, -v[2], v[1],
+        v[2], 0.0, -v[0],
+        -v[1], v[0], 0.0,
+    ])
+
+
+'''
+Template controllers for spin stabilizing and sun pointing.
+'''
+class SpinStabilizingController(ABC):
```
Comment (Ibrassow) [inline flight/apps/adcs/mcm.py:32]: same comment for abstract stuff

**Review by Ibrassow — CHANGES_REQUESTED** [review]
overall, I'm curious about the total memory footprint if yo could measure that but looks good overall

File: flight/apps/adcs/mcm.py [hunk 2]
```diff
@@ -1,10 +1,140 @@
...
+    ang_vel_target = np.linalg.norm(ang_vel_reference)
+
+    @classmethod
+    def update_max_dipole_moment(self) -> None:
+        u_max = np.zeros(3)
+        for row in MagneticCoilAllocator.mat:
+            if not np.all(row == 0.0):
+                row_norm = np.linalg.norm(row)
+                u_max += row / row_norm * (MagnetorquerConst.V_MAX / MagnetorquerConst.V_CONVERT)
+        self.u_max = u_max
```
Comment (Herpderk) [inline flight/apps/adcs/mcm.py:72]: I think there's a more efficient way of doing this part than looping through the matrix, but I haven't quite figured out what it is.

Comment (Ibrassow) [reply flight/apps/adcs/mcm.py:75]: one thing already is to precompute (MagnetorquerConst.V_MAX / MagnetorquerConst.V_CONVERT) into a multiplicative constant (division are extremely expensive)

File: flight/apps/adcs/mcm.py [hunk 3]
```diff
@@ -1,10 +1,140 @@
...
+    if spin_axis[np.argmax(np.abs(spin_axis))] < 0:
+        spin_axis = -spin_axis
+
+    # References and targets
+    ang_vel_reference = spin_axis * MCMConst.REF_FACTOR * ModeConst.STABLE_TOL
+    ang_vel_target = np.linalg.norm(ang_vel_reference)
+
+    @classmethod
+    def update_max_dipole_moment(self) -> None:
+        u_max = np.zeros(3)
```
Comment (Ibrassow) [inline flight/apps/adcs/mcm.py:70]: u_max[:] = 0  # Reset values to zero in place

File: flight/apps/adcs/mcm.py (hunk 2 above)
Comment (Ibrassow) [reply flight/apps/adcs/mcm.py:75]: precompute (MagnetorquerConst.V_MAX / MagnetorquerConst.V_CONVERT) into a multiplicative constant, divisions are expensive

Comment (Ibrassow) [reply flight/apps/adcs/mcm.py:75]: might be worthwile for us to measure the processing speed of this (using the nanosec clock)

File: tests/test_mcm.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    )
+    assert dipole_moment.shape == (3,)
+
+
+def test_sun_pointing(readings) -> None:
+    dipole_moment = get_sun_pointing_dipole_moment(
+        readings["sun_vector"],
+        readings["pointing_error"],
+    )
+    assert dipole_moment.shape == (3,)
```
Comment (Ibrassow) [inline tests/test_mcm.py:42]: Here, test the shape output is good but you want to test the computation as well with a clear input value (instead of random) and compare it to your reference output to make sure the computation is correct. 

You want also to test nominal cases and corner cases (near zero) or when a zero norm appears somewhere.

Be paranoid

**Review by Ibrassow — CHANGES_REQUESTED** [review]

File: flight/apps/adcs/mcm.py (hunk 3 above)
Comment (Herpderk) [reply flight/apps/adcs/mcm.py:70]: this is initializing a new u_max before assigning it to the class attribute, so there's nothing to reset. Unless you want me to directly update the class attribute?

**Review by Ibrassow — APPROVED** [review]
raising an issue for the tests to be fixed in another PR
